### PR TITLE
Add custom algorithm of harmony labeling

### PR DESCRIPTION
**Pull request by yuliazhelt**

### CODE REVIEW

File: harmony_labeling/kp_corpus_process.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
+import numpy as np
+from dataset import add_piece_to_dataset_roots, add_piece_to_dataset_modes
+
+
+def str_time_to_int(timestamp):
+    return int(timestamp.replace('.', ''))
+
+
+def prepare_dataset_roots():
+    with open('/data/datasets/kp-corpus-files/kp-chord-list', 'r') as file:
```
Comment (evgenyim) [inline harmony_labeling/kp_corpus_process.py:10]: Может быть имеет смысл вынести пути в параметры, чтобы их было легче поменять если что

File: harmony_labeling/dynamics.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        probs[root] = major_prob[root] * root_probs_vec[root]
+        probs[root + 12] = minor_prob[root] * root_probs_vec[root]
+    return probs
+
+
+def prob_root_notes(model_roots, model_mode, time_shifts, tokens, i, j):
+    piece = make_piece_from_tokens(time_shifts, tokens, i, j)
+    root_probs_vec = root_probs(model_roots, piece)
+    probs = root_mode_probs(model_mode, piece, root_probs_vec)
+    return np.log(np.max(probs)), np.argwhere(probs == np.max(probs)).flatten(), probs
```
Comment (kliffeup) [inline harmony_labeling/dynamics.py:119]: Можно заранее посчитать max, а затем использовать

File: harmony_labeling/dynamics.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+    for i, (_, ts_time) in enumerate(time_shifts):
+        next_token_i = i
+        next_ts_time = time_shifts[next_token_i][1]
+        while next_ts_time - ts_time <= win_max:
+            if next_ts_time - ts_time < win_min:
+                next_token_i += 1
+                if next_token_i >= len(time_shifts):
+                    break
+                next_ts_time = time_shifts[next_token_i][1]
+                continue
```
Comment (kliffeup) [inline harmony_labeling/dynamics.py:143]: `continue` ничего не делает здесь

File: harmony_labeling/dynamics.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        prev_token_i = j - 1
+        prev_ts_time = time_shifts[prev_token_i][1]
+        
+        while ts_time - prev_ts_time <= win_max and prev_token_i >= 0:
+            if dp[prev_token_i + 1][j + 1] and pref_0[prev_token_i + 1] + dp[prev_token_i + 1][j + 1][0] > pref_0[j + 1]:
+                pref_0[j + 1] = pref_0[prev_token_i + 1] + dp[prev_token_i + 1][j + 1][0]
+                pref_1[j + 1] = prev_token_i
+            prev_token_i -= 1
+            prev_ts_time = time_shifts[prev_token_i][1]
+    return pref_0, pref_1
```
Comment (kliffeup) [inline harmony_labeling/dynamics.py:175]: `pref_0`, `pref_1` поменять бы на что-то более говорящее

File: harmony_labeling/dynamics.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    r = optimal_pieces_inds[curr_step]
+    res = []
+    
+    while curr_step < len(optimal_pieces_inds) - 1:
+        if dp[l + 1][r + 1]:
+            harmony = dp[l + 1][r + 1][1][0]
+            harmonies.append(harmony)
+            time = (time_shifts[l][0], time_shifts[r][0])
+            times.append(time)
+            res.append([int(time[0]), int(time[1]), int(harmony)])
```
Comment (kliffeup) [inline harmony_labeling/dynamics.py:203]: А если dp здесь не отработала, то мы куску ничего не присваиваем? Или такого не бывает?

File: harmony_labeling/kp_corpus_process.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            curr_song_name = line.replace('% ', '').strip('.')
+            song_seq_from_file_dict[curr_song_name] = []
+            
+        elif line[0] != '-':
+            elems = line.strip().replace('  ', ' ').split(' ')
+            song_seq_from_file_dict[curr_song_name].append((str_time_to_int(elems[0]), str_time_to_int(elems[1]), int(elems[-1])))
+            
+    file_notes_dict = dict() # from Melisma notes analyzer
+
+    for filename in song_seq_from_file_dict.keys():
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:28]: `.keys()` при итерировании по словарю обычно не используют

File: harmony_labeling/kp_corpus_process.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        file_notes_dict[filename] = []
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:29]: Чтобы не прописывать такое, обычно используют [defaultdict](https://docs-python.ru/standart-library/modul-collections-python/klass-defaultdict-modulja-collections/)

File: harmony_labeling/kp_corpus_process.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        with open(f'/data/datasets/kp-corpus-files/kp-nbck/{filename}.nc', 'r') as file:
+            lines = file.read().split('\n')
+            for line in lines:
+                if line.split(' ')[0] == 'Note':
+                    file_notes_dict[filename].append((int(line.split(' ')[1]), int(line.split(' ')[2]), int(line.split(' ')[3])))
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:34]: Мб тут заранее сделать split, и потом использовать?

File: harmony_labeling/kp_corpus_process.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            pieces_with_chords.append((curr_notes, end_time - start_time, chord))
+
+    X = []
+    y = []
+    for piece in pieces_with_chords:
+        add_piece_to_dataset_roots(piece, X, y)
+
+    X = np.array(X)
+    y = np.array(y)
+    return X, y
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:53]: Аналогичное замечание про название переменных

File: harmony_labeling/kp_corpus_process.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def prepare_dataset_mode():
+    with open('/data/datasets/kp-corpus-files/kp-chord-list-modes', 'r') as file:
+        lines = file.read().split('\n')
+
+    song_seq_roots_modes_dict = dict() # read file with modes and chord roots
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:60]: Тоже defaultdict можно использовать

File: harmony_labeling/kp_corpus_process.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+            song_seq_roots_modes_dict[curr_song_name] = []
+            
+        elif line[0] != '-':
+            elems = line.strip().replace('  ', ' ').split(' ')
+            song_seq_roots_modes_dict[curr_song_name].append((str_time_to_int(elems[0]), str_time_to_int(elems[1]), int(elems[-2]),  int(elems[-1])))
+
+    file_notes_dict = dict() # from Melisma notes analyzer
+
+    for filename in song_seq_roots_modes_dict.keys():
+        file_notes_dict[filename] = []
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:76]: Аналогично про defaultdict

File: harmony_labeling/kp_corpus_process.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        with open(f'/data/datasets/kp-corpus-files/kp-nbck/{filename}.nc', 'r') as file:
+            lines = file.read().split('\n')
+            for line in lines:
+                if line.split(' ')[0] == 'Note':
+                    file_notes_dict[filename].append((int(line.split(' ')[1]), int(line.split(' ')[2]), int(line.split(' ')[3])))
+    
+    pieces_with_roots_modes= [] # [..., (note, dur), ... ], piece_len, root (one of 12), mode (1 -- major, 2 -- minor)
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:83]: `pieces_with_roots_modes =`

File: harmony_labeling/kp_corpus_process.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+    X = []
+    y = []
+
+    for piece in pieces_with_roots_modes:
+        add_piece_to_dataset_modes(piece, X, y) 
+
+    X = np.array(X)
+    y = np.array(y)
+    return X, y
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:101]: Функция очень похожа на ту, что для корня, мб можно как-то переписать без дублирования кода?

File: harmony_labeling/main.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+import os
+from typing import List
+import coloredlogs
+import argparse
+import json
+import logging
+import tqdm
+from multiprocessing import Pool, context
+from dynamics import harmony_seq_pipeline
+from kp_corpus_process import prepare_dataset_roots, prepare_dataset_mode
```
Comment (kliffeup) [inline harmony_labeling/main.py:11]: Порядок импортов бы поменять, согласно [pep8](https://stackoverflow.com/a/20762915)

File: harmony_labeling/main.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+    'n_estimators': 290,
+    'learning_rate': 0.04774974078860072,
+    'depth': 10,
+    'l2_leaf_reg': 0.10285856054142818,
+    'bootstrap_type': 'Bayesian',
+    'random_strength': 0.0025811780488595813,
+    'bagging_temperature': 0.13879907325154095,
+    'od_type': 'IncToDec',
+    'od_wait': 28
+}
```
Comment (kliffeup) [inline harmony_labeling/main.py:36]: Конфиги лучше вынести в json'ы

File: harmony_labeling/main.py [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+def get_args(default='.') -> argparse.Namespace:
```
Comment (kliffeup) [inline harmony_labeling/main.py:38]: Лучше дефолтный путь назвать `default_path`

File: harmony_labeling/main.py [hunk 17]
```diff
@@ -0,0 +1,209 @@
...
+                    break
+    pbar.close()
+
+    for r in res:
+
+        file_name, new_output_folder, base_name, d = r
+
+        if d is not None and len(d) != 0:
+            files_result['{0}/{1}'.format(new_output_folder, base_name)] = []
+            files_result[
```
Comment (kliffeup) [inline harmony_labeling/main.py:199]: Лучше переменную files_results объявить поближе к этой части кода

Comment (kliffeup) [reply harmony_labeling/main.py:199]: И тоже через defaultdict

File: harmony_labeling/main.py [hunk 18]
```diff
@@ -0,0 +1,209 @@
...
+
+
+    model_roots = CatBoostClassifier(**best_params_roots, logging_level='Silent')
+    X_roots, y_roots = prepare_dataset_roots()
+    model_roots.fit(X_roots, y_roots)
+
+
+    model_mode = CatBoostClassifier(**best_params_mode, logging_level='Silent')
+    X_mode, y_mode = prepare_dataset_mode()
+    model_mode.fit(X_mode, y_mode)
```
Comment (kliffeup) [inline harmony_labeling/main.py:153]: Модели каждый раз переобучаются? Лучше бы вынести два процесса в разные функции, файлы: в одном код для переобучения модели, в другом -- уже запуск их для инференса.

File: harmony_labeling/main.py (hunk 15 above)
Comment (kliffeup) [reply harmony_labeling/main.py:36]: И конфиги здесь не фиксируют seed для обучающихся моделей. Мне кажется, Catboost не фиксирует рандом по умолчанию, так что для дальнейшего кода для переобучения эти параметры могут выдавать разные результаты при инференсе для разных запусков кода. Так что лучше сохранить состояние этих моделей и подгружать их для инференса, чем как это сделано сейчас

File: harmony_labeling/kp_corpus_process.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+    curr_song_name = ''
+
+    for line in lines:
+        if line[0] == '%':
+            curr_song_name = line.replace('% ', '').strip('.')
+            song_seq_from_file_dict[curr_song_name] = []
+            
+        elif line[0] != '-':
+            elems = line.strip().replace('  ', ' ').split(' ')
+            song_seq_from_file_dict[curr_song_name].append((str_time_to_int(elems[0]), str_time_to_int(elems[1]), int(elems[-1])))
```
Comment (kliffeup) [inline harmony_labeling/kp_corpus_process.py:24]: То есть выполняется условие `elif`а, то `curr_song_name` не поменяется здесь? Это ожидаемо
